example/quadratic.f90

```diff
@@ -0,0 +1,151 @@
+
+  ! Calculate predictions on the test set
+  ypred = [(net % predict([xtest(i)]), i = 1, test_size)]
+
+  ! Print the mean squared error
+  print '(i0,1x,f9.6)', num_iterations, sum((ypred - ytest)**2) / size(ypred)
+
+  deallocate(x, y, xtest, ytest, ypred)
+end program quadratic_fit
+
+
+subroutine sgd_optimizer(net, x, y, learning_rate, num_iterations)
+  type(network), intent(inout) :: net
+  real, dimension(:), intent(in) :: x, y
+  real, intent(in) :: learning_rate
+  integer, intent(in) :: num_iterations
+  integer :: i, n, num_samples
+
+  num_samples = size(x)
+
+  do n = 1, num_iterations
+    do i = 1, num_samples
+      call net % forward([x(i)])
+      call net % backward([y(i)])
+      ! SGD update
+      call update_parameters(net, learning_rate)
+    end do
+  end do
+end subroutine sgd_optimizer
+
+
+subroutine batch_sgd_optimizer(net, x, y, learning_rate, num_iterations)
+  type(network), intent(inout) :: net
+  real, dimension(:), intent(in) :: x, y
+  real, intent(in) :: learning_rate
+  integer, intent(in) :: num_iterations
+  integer :: i
+
+  call net % forward(x)
+  call net % backward(y)
+  ! Batch SGD update
+  call update_parameters(net, learning_rate)
+
+  do i = 2, num_iterations
+    call net % forward(x)
+    call net % backward(y)
+    ! Accumulating gradients
+    call accumulate_gradients(net)
+  end do
+
+  ! Updating parameters
+  call update_parameters(net, learning_rate)
+end subroutine batch_sgd_optimizer
+
+
+subroutine minibatch_sgd_optimizer(net, x, y, learning_rate, num_iterations, batch_size)
+  type(network), intent(inout) :: net
+  real, dimension(:), intent(in) :: x, y
+  real, intent(in) :: learning_rate
+  integer, intent(in) :: num_iterations, batch_size
+  integer :: i, n, num_samples, num_batches, start_index, end_index
+  real, dimension(:), allocatable :: batch_x, batch_y
+
+  num_samples = size(x)
+  num_batches = num_samples / batch_size
+
+  allocate(batch_x(batch_size), batch_y(batch_size))
+
+  do n = 1, num_iterations
+    do i = 1, num_batches
+      ! Selecting batch
+      start_index = (i - 1) * batch_size + 1
+      end_index = i * batch_size
+      batch_x = x(start_index:end_index)
+      batch_y = y(start_index:end_index)
+
+      call net % forward(batch_x)
+      call net % backward(batch_y)
+      ! Mini-batch SGD update
+      call update_parameters(net, learning_rate)
+    end do
+  end do
+
+  deallocate(batch_x, batch_y)
+end subroutine minibatch_sgd_optimizer
+
+
+! subroutine update_parameters(net, learning_rate)
+!   type(network), intent(inout) :: net
+!   real, intent(in) :: learning_rate
+!   integer :: i, num_layers
+!   type(dense_layer) :: layer
+
+!   num_layers = net % num_layers()
+
+!   do i = 1, num_layers
+
+!     layer = net % get_layer(i)
+!     layer % weights = layer % weights - learning_rate * layer % gradients
+!     layer % biases = layer % biases - learning_rate * layer % gradients
+!   end do
+! end subroutine update_parameters
```

Comment:
I don't think you'll need a custom subroutine for updating here.

I think you can just use the `network % update` method as in

https://github.com/modern-fortran/neural-fortran/blob/8293118220ec1e9c16abda766e3b273f161592b4/src/nf/nf_network_submodule.f90#L553

and in your subroutines that would probably be

```fortran
call net % update(learning_rate / batch_size)
```